kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -233,48 +288,148 @@ class KinesisSink private (
+            sendToSqsOrRetryToKinesis(failures, nextBackoff)(retryErrorMsg)
           }
         }
         case Failure(f) => {
           log.error("Writing failed.", f)
-          log.error(s"Retrying in $nextBackoff milliseconds...")
-          scheduleBatch(batch, nextBackoff)
+          val retryErrorMsg = s"Retrying in $nextBackoff milliseconds..."
+          sendToSqsOrRetryToKinesis(batch, nextBackoff)(retryErrorMsg)
         }
       }
     }
 
-  private def multiPut(name: String, batch: List[(ByteBuffer, String)]): Future[PutRecordsResult] =
+  private def sendToSqsOrRetryToKinesis(
+    failures: List[BatchEntry],
+    nextBackoff: Long
+  )(
+    retryErrorMsg: String
+  ): Unit =
+    maybeSqs match {
+      case Some(sqs) =>
+        log.info(
+          s"Sending ${failures.size} events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+        )
+        putToSqs(sqs, failures)
+        ()
+      case None =>
+        log.error(retryErrorMsg)
+        log.warn(
+          s"${failures.size} failed events scheduled for retry (consider setting a SQS Buffer in config.hocon)"
+        )
+        scheduleBatch(failures, nextBackoff)
+    }
+
+  private def putToSqs(sqs: SqsClientAndName, batch: List[BatchEntry]): Future[Unit] =
+    Future {
+      log.info(s"Writing ${batch.size} messages to SQS queue: ${sqs.sqsBufferName}")
+      val sqsBatchEntries = batch.map(addUuids).map(toSqsBatchEntry)
+      val MaxSqsBatchSize = 10
+      sqsBatchEntries.grouped(MaxSqsBatchSize).foreach { batchEntryGroup =>
+        sendToSqsWithRetry(sqs, batchEntryGroup)
+      }
+    }
+
+  /**
+   *  UUIDs are generated for retry purpose. When batch fails, we want to retry (limited number of times)
+   *   only those entries that failed within the batch.
+   */
+  private val addUuids: (BatchEntry) => (BatchEntryWithId) = {
+    case BatchEntry(msg, key) => BatchEntryWithId(msg, key, java.util.UUID.randomUUID())
+  }
+
+  private val toSqsBatchEntry: (BatchEntryWithId) => (UUID, SendMessageBatchRequestEntry) = {
+    case BatchEntryWithId(msg, key, id) =>
+      val b64EncodedMsg = encode(msg)
+      val batchReqEntry = new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64EncodedMsg)
+        .withMessageAttributes(
+          Map(
+            "kinesisKey" ->
+              new MessageAttributeValue()
+                .withDataType("String")
+                .withStringValue(key)
+          ).asJava
+        )
+        .withId(id.toString)
+      (id, batchReqEntry)
+  }
+
+  private def createBatchRequest(queueUrl: String, batch: List[SendMessageBatchRequestEntry]) =
+    new SendMessageBatchRequest()
+      .withQueueUrl(queueUrl)
+      .withEntries(batch.asJava)
+
+  private def sendToSqsWithRetry(
+    sqs: SqsClientAndName,
+    batchEntryGroup: List[(UUID, SendMessageBatchRequestEntry)]
+  ) = {
+    val current = TrieMap(batchEntryGroup: _*)
```

Comment:
Given that SQS is to be used when there is a surged and collector is stressed, it would be nice to not need to copy all the data that we want to send each time.

Reply:
It would be nice, that's true. Tests are showing that cpu utilization is less than it was before. Of course we would see the real numbers on production. The solution here is not the prettiest, but it lets us send batches of events instead of sending one event per api call. We might improve it in the future though. Unless you have some good ideas how to make it readable and not copy the list?